### PR TITLE
Conserve memory setting backport to 5.0.

### PR DESCRIPTION
**Pull request by PeterSolMS**
This is the first snapshot of the port. I'll run a few sanity checks on it tomorrow.

### CODE REVIEW

File: src/coreclr/src/gc/gc.cpp [hunk 1]
```diff
@@ -31998,6 +32053,19 @@ size_t gc_heap::desired_new_allocation (dynamic_data* dd,
...
 
             f = surv_to_growth (cst, limit, max_limit);
+            if (conserve_mem_setting != 0)
+            {
+                // if this is set, compute a growth factor based on it.
+                // example: a setting of 6 means we have a goal of 60% live data
+                // this means we allow 40% fragmentation
+                // to keep heap size stable, we only use half of that (20%) for new allocation
+                // f is (live data + new allocation)/(live data), so would be (60% + 20%) / 60% or 1.33
+                float f_conserve = (10.0f / conserve_mem_setting - 1) * 0.5f + 1.0f;
```
Comment (Maoni0) [inline src/coreclr/src/gc/gc.cpp:32063]: nit - ()

                `float f_conserve = ((10.0f / conserve_mem_setting) - 1) * 0.5f + 1.0f;`

this is what you had for framework :D

**Review by Maoni0 — APPROVED** [review]
LGTM other than the minor formatting issue.

File: src/coreclr/src/gc/gc.cpp (hunk 1 above)
Comment (PeterSolMS) [reply src/coreclr/src/gc/gc.cpp:32063]: I see - added the extra pair of parentheses...

**Comment by PeterSolMS** [comment]
**Description**
This adds a GC config setting to favor conserving memory over maximizing performance. This can be set via the environment variable COMPLUS_GcConserveMemory, with values between 1 and 9, e.g.: "set COMPLUS_GcConserveMemory= 5". Intuitively, the higher the value, the harder GC tries to reduce memory footprint, at the cost of more frequent GCs and possibly higher pause time. It's a port of a feature we introduced for .NET Framework.
 
**Customer Impact**
This allows customers to reduce memory footprint and possibly put more managed processes on machines.

**Regression?**
No.

**Risk (see taxonomy)**
Low. There is no behavior change if the environment variable is not set.

**Link the PR to the original issue and/or the PR to master.**
[https://github.com/dotnet/runtime/pull/56772](url)

**Is there a packaging impact?**
No.

**Comment by hoyosjs** [comment]
The failures are well known, don't block on those. #56295 and #55271

@karelz do you know if we can disable these in 5.0 to allow for servicing fixes?

**Comment by jeffschwMSFT** [comment]
@leecow I want to double check the milestone.  This is part of 5.0.10, and not 5.0.11, correct?

**Comment by leecow** [comment]
Yes, sorry for the confusion - Anirudh's merge happened early enough to be included in the September update (5.0.10).

**Pull request merged**